src/lib/bcmp/bcmp.cpp

```diff
@@ -484,3 +556,124 @@ void bcmp_init(struct netif* netif, NvmPartition * dfu_partition, Configuration*
                      NULL);
   configASSERT(rval == pdPASS);
 }
+
+
+static bool _message_list_add_message(bcmp_request_element *message) {
+  bool rval = false;
+  configASSERT(message);
+  if (xSemaphoreTake(_ctx.messages_list_mutex, pdMS_TO_TICKS(DEFAULT_MESSAGE_TIMEOUT_MS)) == pdPASS) {
+    if (_ctx.messages_list == NULL) {
+      _ctx.messages_list = message;
+    } else {
+      bcmp_request_element *current = _ctx.messages_list;
+      while (current->next != NULL) {
+        current = current->next;
+      }
+      current->next = message;
+    }
+    rval = true;
+    xSemaphoreGive(_ctx.messages_list_mutex);
+  }
+  return rval;
+}
+
+// NOTE: This function is not thread safe and can only be called within a section
+// protected by the messages_list_mutex semaphore.
+static bool _message_list_remove_message(bcmp_request_element *message) {
```

Comment:
The other message list modification functions manage the semaphore internally themselves. Why do we require the caller to manage the semaphore for this one?

Reply:
`_message_list_remove_message` is called from a loop within `_message_list_timer_expiry_cb` where we have already taken the mutex.

Reply:
Ah, got it. I see that. Thanks, and yeah, we want to remove all expired messages in one pass rather than potentially having to wait for the semaphore in the middle of the loop. 👍